### PR TITLE
*: support backup/restore with  NFS #148

### LINKED ISSUE #148: [feature] Backup to nfs server , restore from nfs server
**Is your feature request related to a problem? Please describe.**

**Describe the solution you'd like**
Make a backup like this:
add feild BackupToNFS specify the nfs server in yaml file, then
```shell
kubectl apply -f mysql_v1alpha1_backup.yaml 
```
Can backup to Nfs server, 
add feild RestoreFromNFS specify the nfs server in yaml file, then
 it can restore cluster by:
```shell
kubectl apply -f mysql_v1alpha1_cluster.yaml 
```
Just add 

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by acekingke**
### What type of PR is this?

/enhancement
### Which issue(s) this PR fixes?

Fixes #148 

### What this PR does?

Summary:
1. create nfs resouce and nfs service
2. support backup to nfs and restore from nfs

### CODE REVIEW

File: config/samples/nfs_server.yaml [hunk 1]
```diff
@@ -0,0 +1,14 @@
...
+spec:
+  ports:
+    - name: nfs
+      port: 2049
+    - name: mountd
+      port: 20048
+    - name: rpcbind
+      port: 111
+  selector:
+    role: nfs-server
```
Comment (runkecheng) [inline config/samples/nfs_server.yaml:14]: Blank line at the end.

**Comment by andyli029** [comment]
staticcheck Failed.

File: sidecar/init.go [hunk 2]
```diff
@@ -214,10 +214,25 @@ func runInitCommand(cfg *Config) error {
 			var err_f error
 			if cfg.CloneFlag {
 				err_f = cfg.executeCloneRestore()
-				if err_f != nil {
-					return fmt.Errorf("failed to execute Clone Restore : %s", err_f)
-				}
+
 			} else {
+
```
Comment (andyli029) [inline sidecar/init.go:219]: Delete the space line.

File: api/v1alpha1/backup_types.go [hunk 3]
```diff
@@ -35,7 +35,12 @@ type BackupSpec struct {
 	Image string `json:"image"`
 
 	// HostName represents the host for which to take backup
-	HostName string `json:"hostname"`
+	// If is empty, is use leader HostName
+	HostName string `json:"hostname,omitempty"`
+
+	// Represents the name of backup to NFS
```
Comment (andyli029) [inline api/v1alpha1/backup_types.go:41]: Is `IP of NFS server`?

File: backup/syncer/job.go [hunk 4]
```diff
@@ -115,10 +115,43 @@ func (s *jobSyncer) ensurePodSpec(in corev1.PodSpec) corev1.PodSpec {
 	in.RestartPolicy = corev1.RestartPolicyNever
 	sctName := fmt.Sprintf("%s-secret", s.backup.Spec.ClusterName)
 	in.Containers[0].Name = utils.ContainerBackupName
+
```
Comment (andyli029) [inline backup/syncer/job.go:118]: Delete it.

File: backup/syncer/job.go [hunk 5]
```diff
@@ -115,10 +115,43 @@ func (s *jobSyncer) ensurePodSpec(in corev1.PodSpec) corev1.PodSpec {
...
+				},
+			},
+		}
+		//"rm -rf /backup/*;curl --user sys_backups:sys_backups sample-mysql-0.sample-mysql.default:8082/download|xbstream -x -C /backup"
+		in.Containers[0].Command = []string{
+			"/bin/bash", "-c", "--",
+		}
+		var backupToDir string = utils.BuildBackupName()
+		in.Containers[0].Args = []string{
+			fmt.Sprintf("mkdir -p /backup/%s;curl --user $BACKUP_USER:$BACKUP_PASSWORD %s/download|xbstream -x -C /backup/%s; exit ${PIPESTATUS[0]}",
```
Comment (andyli029) [inline backup/syncer/job.go:140]: ->
```
			fmt.Sprintf("mkdir -p /backup/%s; " +
				"curl --user $BACKUP_USER:$BACKUP_PASSWORD %s/download|xbstream -x -C /backup/%s; " +
				"exit ${PIPESTATUS[0]}",
				backupToDir, 
				s.backup.GetBackupURL(s.backup.Spec.ClusterName, s.backup.Spec.HostName), backupToDir),
```

**Review by andyli029 — COMMENTED** [review]
Add subdir `back-nfs` in the samples about and mv related file into it.

File: config/samples/nfs_rc.yaml [hunk 6]
```diff
@@ -0,0 +1,32 @@
...
+            containerPort: 111
+        securityContext:
+          privileged: true
+        volumeMounts:
+          - mountPath: /exports
+            name: nfs-export-fast
+      volumes:
+        - name: nfs-export-fast
+          persistentVolumeClaim:
+            claimName: backup-pv-volume
```
Comment (runkecheng) [inline config/samples/nfs_rc.yaml:32]: claimName should be pvc name: backup-pv-claim

File: config/samples/pv-volume.yaml [hunk 7]
```diff
@@ -0,0 +1,14 @@
+apiVersion: v1
```
Comment (runkecheng) [inline config/samples/pv-volume.yaml:1]: 1. add sc
2. support source is LocalVolume?

**Review by runkecheng — CHANGES_REQUESTED** [review]

File: docs/deploy_backup_restore_nfs.md [hunk 8]
```diff
@@ -0,0 +1,60 @@
+# mysql-operator
+
+## Quickstart for  NFS backup
+### Create NFS server
+First create your PVC, such as "neosan1", 
+Or use local storage pvc, see `config/sample/pv-claim.yalm`
```
Comment (andyli029) [inline docs/deploy_backup_restore_nfs.md:6]: yalm
->
yaml?

**Review by runkecheng — CHANGES_REQUESTED** [review]
The name of the backup folder cannot distinguish the backup of multiple clusters, Change the name such as `<cluster_name>_202241564422`.
`
sh-4.2# ls exports/
backup_202241564422  index.html
sh-4.2# ls exports/backup_202241564422/
backup-my.cnf   ibdata1  performance_schema  sys                     xtrabackup_checkpoints  xtrabackup_logfile
ib_buffer_pool  mysql    radondb             xtrabackup_binlog_info  xtrabackup_info
`

**Comment by acekingke** [comment]
https://github.com/radondb/radondb-mysql-kubernetes/pull/232#discussion_r851094310 
cannot use GetNameForResource, it is `MysqlCluster` method, can not call in sidecar

File: config/samples/pv-volume.yaml [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+spec:
+  storageClassName: manual
+  nodeAffinity:
+    required:
+      nodeSelectorTerms:
+      - matchExpressions:
+        - key: kubernetes.io/hostname
+          operator: In
+          values:
+          - node2
```
Comment (runkecheng) [inline config/samples/pv-volume.yaml:26]: nodeAffinity is not necessary， if haven't node2, PV will not be created.

File: docs/deploy_backup_restore_nfs.md [hunk 10]
```diff
@@ -0,0 +1,60 @@
...
+if create the nfs server successful, get the then:
+
+## config `mysql_v1alpha1_backup.yaml ` and backup
+Add field in `mysql_v1alpha1_backup.yaml ` as follow:
+```yaml
+BackupToNFS: "IP of NFS server"
+```
+use command as follow to backup
+```shell
+kubectl apply -f config/samples/mysql_v1alpha1_backup.yaml
```
Comment (runkecheng) [inline docs/deploy_backup_restore_nfs.md:33]: Notice: backup cr and mysqlcluster cr must be in the same namespace.

**Comment by acekingke** [comment]
> The name of the backup folder cannot distinguish the backup of multiple clusters, Change the name such as <cluster_name>_202241564422.



it has been modified

File: config/samples/mysql_v1alpha1_backup.yaml [hunk 11]
```diff
@@ -5,5 +5,7 @@ metadata:
 spec:
   # Add fields here
   image: radondb/mysql-sidecar:latest
```
Comment (runkecheng) [inline config/samples/mysql_v1alpha1_backup.yaml:7]: mysql57-sidecar:v2.2.0

**Review by runkecheng — APPROVED** [review]

**Review by andyli029 — APPROVED** [review]
Merge first, optimize further.

**Pull request merged**